core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1966,29 +1966,78 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      fibers: Iterable[Fiber[E, _]],
+      promise: Promise[E, List[B]],
+      buffer: Array[B],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.halt(e).unit *> Fiber.interruptAll(fibers),
+        a =>
+          todo.modify { t =>
+            buffer.update(idx, a)
+            if (t == 1) promise.succeed(buffer.toList).unit -> 0 else UIO.unit -> (t - 1)
+          }.flatten
+      )
+
+    (for {
+      size    <- UIO.effectTotal(as.size)
+      todo    <- Ref.make(size)
+      buffer  <- UIO.effectTotal(new Array[Any](size).asInstanceOf[Array[B]])
```

Comment:
```suggestion
      buffer  <- UIO.effectTotal(Array.ofDim[B](size))
```

Reply:
Also, it's probably best to move the buffer into a `Ref`.

Reply:
What is gain of keeping *mutable* array in `Ref`? I think that reference will never change, so `Ref#modify` will never discover that value was changed.
Secondly - we know that each fiber operates only on it's index, so that are really independent operations.
What can go wrong?

Re: `ofDim` - I understand this gives better performance for primitive types but requires `ClassTag[B]` - do we want to change method signatures?